Public/Src/Tools/Tool.MsBuildGraphBuilder/PropertyTrackingLogger.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using BuildXL.Utilities.Collections;
+using Microsoft.Build.Framework;
+
+namespace ProjectGraphBuilder
+{
+    internal class PropertyTrackingLogger : ILogger
+    {
+        private readonly HashSet<string> _variablesRead = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

```

Comment:
Why the locking and not simply a `ConcurrentSet<string>`?

Reply:
If we keep locking, we have the pattern to create a seperate object for the lock and are not locking on the set directly. The fact that you don't expose the set makes this case safe, but it is too easy to accidentally leak it and expose issues later.

i.e. `private readonly object m_variablesReadLock =new object()`

---
In reply to: [306899167](https://github.com/microsoft/BuildXL/pull/648#discussion_r306899167) [](ancestors = 306899167)

Reply:
@dannyvv, I'm looking for a ConcurrentSet in the codebase and the only thing I see is:
[ConcurrentHashSet](https://github.com/microsoft/BuildXL/blob/master/Private/Wdg/ExecutionLogSdk/ConcurrentHashSet.cs) and [ConcurentBigSet](https://github.com/microsoft/BuildXL/blob/master/Public/Src/Utilities/Collections/ConcurrentBigSet.cs)

The former is in a project not referenced by this one. The latter is geared towards large sets where in this instance it will almost always be <1000 elements.

Am I missing something?

Should I a) bring over the ConcurrentHashSet class above into the current project, duplicating code, b) bring the ConcurrentHashSet class into the Utilities project to be referenced by both the project currently hosting that class and this one? c) Use a standard class like ConcurrentDictionary<string, object> and essentially use it as a hash set by disregarding the value?  I'd think option B to be the best, but I'm certainly not sure of the ramifications of moving such classes around.

Thoughts?

Reply:
Ah didn't think about referneces....
My personal preference would then be use ConcurrentDictionary<T,U>

---
In reply to: [308304479](https://github.com/microsoft/BuildXL/pull/648#discussion_r308304479) [](ancestors = 308304479)

Reply:
Changed to ConcurrentDictionary in latest commit.